File: labour_service/src/user/domain/repository.py
from typing import Protocol

from src.user.domain.entity import User
from src.user.domain.value_objects.user_id import UserId


class UserRepository(Protocol):
    """Repository interface for User entity"""

    async def save(self, user: User) -> None:
        """
        Save or update a user.

        Args:
            user: The user to save
        """

    async def delete(self, user: User) -> None:
        """
        Delete a user.

        Args:
            user: The user to delete
        """

    async def get_by_id(self, user_id: UserId) -> User | None:
        """
        Retrieve a user by their ID.

        Args:
            user_id: The ID of the user to retrieve

        Returns:
            The user if found, else returns None
        """

    async def get_by_ids(self, user_ids: list[UserId]) -> list[User]:
        """
        Retrieve a list of users by their IDs.

        Args:
            user_ids: The IDs of the users to retrieve

        Returns:
            A list of users
        """

File: services/user-service/src/domain/repository.rs
use super::entity::User;
use super::exceptions::RepositoryError;
use super::value_objects::UserId;

use async_trait::async_trait;

#[async_trait(?Send)]
pub trait UserRepository: Send + Sync {
    async fn save(&self, user: &User) -> Result<(), RepositoryError>;

    async fn get_by_id(&self, user_id: &UserId) -> Result<Option<User>, RepositoryError>;

    async fn get_by_email(&self, email: &str) -> Result<Option<User>, RepositoryError>;

    async fn exists(&self, user_id: &UserId) -> Result<bool, RepositoryError>;

    async fn delete(&self, user_id: &UserId) -> Result<(), RepositoryError>;
}
